### DISCUSSION TITLE
are `@react-aria/*` packages more than "accessible primitives" now?

### DISCUSSION THREAD

**Discussion by csantos1113** [adobe/react-spectrum#5458] [General]
after seeing the work done in #5288 that introduced form validation to the `@react-aria` packages I have a couple clarifying questions regarding the intention of `@react-aria/*` pkgs and specifically about form integration with external libraries like [react-hook-form](https://github.com/react-hook-form/react-hook-form)

### 1️⃣ Are `@react-aria/*` hooks doing more than what they were originally intended to do?

from https://react-spectrum.adobe.com/react-aria/:

> _A library of React Hooks that provides accessible UI primitives for your design system._

I've used these hooks as low-level elements that give me the correct accessible props as well as event handlers for my components to work (aria-*, focus, hover, labelling, etc) where I just passed some props in and these hooks gave me props back for my HTML elements to be accessible.

But specifically talking about `@react-aria/textfield` (`useTextField` hook) (and other fields) there is a lot of logic that I wonder if they are truly part of the "accessible hook primitives", and/or if they should be moved somewhere else:
- https://github.com/adobe/react-spectrum/blob/75417bb15cc7bec44defd2ff71e7d22d56453e85/packages/%40react-aria/textfield/src/useTextField.ts#L139
- https://github.com/adobe/react-spectrum/blob/75417bb15cc7bec44defd2ff71e7d22d56453e85/packages/%40react-aria/textfield/src/useTextField.ts#L140
- https://github.com/adobe/react-spectrum/blob/75417bb15cc7bec44defd2ff71e7d22d56453e85/packages/%40react-aria/textfield/src/useTextField.ts#L142-L158
- https://github.com/adobe/react-spectrum/blob/75417bb15cc7bec44defd2ff71e7d22d56453e85/packages/%40react-aria/textfield/src/useTextField.ts#L207-L208

### 2️⃣ form-validation in `@react-aria/*` plus react-hook-form integration

I see there is a documentation section for how to integrate with react-hook-form https://react-spectrum.adobe.com/react-aria/forms.html#react-hook-form; but after navigating through the code

- https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/form/src/useFormValidation.ts 
- and https://github.com/adobe/react-spectrum/blob/main/packages/%40react-stately/form/src/useFormValidationState.ts

there is a lot of JS running always, regardless if the validation is handled from the outside (in our case, by react-hook-form itself).

It seems to me that we have 2 validation mechanisms running at the same time, so I wonder if there is a way to opt-out of `@react-aria` validation and let the third-party handle the validation for us?

_which piggybacks to my 1️⃣ question: should these primitive hooks stay "dumb" and move that extra logic somewhere else?_

---

I appreciate your help talking through these points and what would be the overall roadmap 🙏

**Comment by devongovett** [comment]
We are certainly adding more features to React Aria to make it easier to use, but I wouldn't consider form validation to be outside its realm. In our case, we are mostly just wrapping the native HTML form validation with a nicer API. If you want to use a separate form library that's also fine – all of the built-in features can be turned off, and they are pretty small so shouldn't affect much. Native validation is nice, but we know that people are mixing our components with others so external form libraries are also supported. As long as it's not getting in the way of anything it seems fine to me.

**Reply by csantos1113** [reply]
Hey @devongovett thanks for replying back. 

Please allow me to dive into this particular topic:`@react-stately/form`. It has a lot of built-in logic that as far as I can tell, I'm not able to turn off.

For example:

- `useFormValidationState` is used by all [form related hooks in `@react-aria/*`](https://github.com/search?q=repo%3Aadobe%2Freact-spectrum+useFormValidationState&type=code)
- that [hook uses a context](https://github.com/adobe/react-spectrum/blob/a2146aa2eb9cc28f4d90c8f520e6be5ff90471c4/packages/%40react-stately/form/src/useFormValidationState.ts#L99)
- but that context is only initialized [in `react-aria-components`](https://github.com/search?q=repo%3Aadobe%2Freact-spectrum+FormValidationContext+path%3A%2F%5Epackages%5C%2Freact-aria-components%5C%2Fsrc%5C%2F%2F&type=code)
  - And I'm not using `react-aria-components`
- Even without that context, react-hook-form is the one doing the validation in my case
  - I use `react-hook-form` + `useTextField`; that's pretty much it

Which is what I meant by _"there is a lot of JS running always, regardless if the validation is handled from the outside"_ and would be great to just not have it running in my render cycles.

BTW: having react-hook-form doing the validation with the latest `@react-aria` works fine as described in https://react-spectrum.adobe.com/react-aria/forms.html#react-hook-form, but that doesn't solve the point that there is extra code running all the time (in this case: `@react-stately/form` wanting to validate the fields).

---

Deep thought here: Looking from the outside, Devon, _could_ it be possible that [react-aria](https://github.com/adobe/react-spectrum/tree/main/packages/react-aria) is unintentionally growing in a direction that is too tied to [react-aria-components](https://github.com/adobe/react-spectrum/tree/main/packages/react-aria-components) instead of being pure primitives?

---

Thanks for your time talking through this!

**Reply by devongovett** [reply]
Is this a theoretical concern or do you have a real-world issue with it? By default when using the hooks, native validation is not enabled. So those hooks won't do anything unless you enable it with `validationBehavior="native"`, or use the server validation or `isInvalid` props (`isInvalid` always existed and is used for integration with other form libraries). These are not specific to React Aria Components, and can be used in other Hook-based implementations as well (we do this in our React Spectrum library). IMO these are useful features that are pretty much table-stakes for any component library, so building it into React Aria simplifies things a lot and avoids the need for other libraries in many cases. But if you have a specific functionality or performance problem that you're having please let us know!

**Reply by csantos1113** [reply]
> So those hooks won't do anything unless you enable it with `validationBehavior="native"`, or use the server validation or `isInvalid` props (`isInvalid` always existed and is used for integration with other form libraries).

While this is true, it's also true the code within the hooks are running always, on every re-render, for each field.

The following lines will always run on every render per field albeit they don't do anything:
https://github.com/adobe/react-spectrum/blob/fc0e6b6587404b7e493d09e814133c0b546b0bd1/packages/%40react-stately/form/src/useFormValidationState.ts#L76-L186

I haven't ran benchmarks, but that's quite a lot of javascript running on every re-render, even if it does nothing.

> (`isInvalid` always existed and is used for integration with other form libraries)

In the ideal scenario, something like this is what would make the integration with other form libraries solid (note: pseudo code, but that's the overall idea):

```tsx
let { isInvalid } = props;

if (typeof isInvalid === 'boolean') {
  return early because it's controlled
}

return useFormValidationStateImpl(props);
```

**Reply by devongovett** [reply]
Unfortunately hooks cannot be conditional in React so that's impossible. I really don't think this should have any measurable performance impact. If you don't pass in a custom `validate` function it isn't run, if there are no server errors on context, nothing happens, if no native validation occurred the effect returns early. The only impact would be two extra useState calls and one useRef, which should be essentially free. If you have real-world benchmarks showing a performance issue we're happy to look into it though.
